### PR TITLE
feat(swarm): make stream uprade errors more ergonomic

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

The currently provided `ConnectionHandlerUpgrErr` is very hard to use. Not only does it have a long name, it also features 3 levels of nesting which results in a lot of boilerplate. Last but not least, it exposes `multistream-select` as a dependency to all protocols.

We fix all of the above by renaming the type to `StreamUpgradeError` and flattening out its interface. Unrecoverable errors during protocol selection are hidden within the `Io` variant.

Related: #3759.

## Notes & open questions

~Draft because it depends on https://github.com/libp2p/rust-libp2p/pull/3605.~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/CHANGELOG.md [hunk 1]
```diff
@@ -14,11 +14,16 @@
...
+- Deprecate `ConnectionHandlerUpgrErr` in favor of `StreamUpgradeError`.
+  This new type features a flat interface which should be easier to use.
+  See [PR XXXX].
+
 [PR 3605]: https://github.com/libp2p/rust-libp2p/pull/3605
 [PR 3746]: https://github.com/libp2p/rust-libp2p/pull/3746
 [PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
 [PR 3746]: https://github.com/libp2p/rust-libp2p/pull/3746
 [PR 3865]: https://github.com/libp2p/rust-libp2p/pull/3865
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline swarm/CHANGELOG.md]:
```suggestion
  See [PR 3882].

[PR 3605]: https://github.com/libp2p/rust-libp2p/pull/3605
[PR 3746]: https://github.com/libp2p/rust-libp2p/pull/3746
[PR 3715]: https://github.com/libp2p/rust-libp2p/pull/3715
[PR 3746]: https://github.com/libp2p/rust-libp2p/pull/3746
[PR 3865]: https://github.com/libp2p/rust-libp2p/pull/3865
[PR 3882]: https://github.com/libp2p/rust-libp2p/pull/3882
```

File: swarm/src/handler.rs [hunk 2]
```diff
@@ -459,54 +459,67 @@ impl<TConnectionUpgrade, TOutboundOpenInfo, TCustom, TErr>
...
-pub enum ConnectionHandlerUpgrErr<TUpgrErr> {
+pub enum StreamUpgradeError<TUpgrErr> {
     /// The opening attempt timed out before the negotiation was fully completed.
     Timeout,
-    /// Error while upgrading the substream to the protocol we want.
-    Upgrade(UpgradeError<TUpgrErr>),
+    /// The upgrade produced an error.
+    Apply(TUpgrErr),
+    /// No protocol could be agreed upon.
+    NegotiationFailed,
```
Comment (mxinden) [inline swarm/src/handler.rs:472]: As a side-note, we had multiple users complain, that it is hard to debug why the negotiation failed. What would have helped them is `NegotiationFailed` to contain the protocols the local and the remote peer offered.

That said, using `DEBUG` logging and looking for the `multistream-select` lines also helps the debugging, though a bit harder to discover.

Not to be addressed in this pull request.

File: swarm/CHANGELOG.md [hunk 3]
```diff
@@ -14,11 +14,16 @@
   This variant was never constructed and thus dead code.
   See [PR 3605].
 
+- Deprecate `ConnectionHandlerUpgrErr` in favor of `StreamUpgradeError`.
+  This new type features a flat interface which should be easier to use.
+  See [PR 3882].
```
Comment (mxinden) [inline swarm/CHANGELOG.md]: Changelog reads like the `ConnectionHandlerUpgrErr` doesn't change. Though given that it is a type alias, it is flattened just like `StreamUpgradeError`.

Would something along the lines below help?
```suggestion
- Flatten `ConnectionHandlerUpgrErr` and rename to `StreamUpgradeError`.
  See [PR 3882].
```

**Review by mxinden — APPROVED** [review]
Change looks good to me. Just one comment.

File: swarm/CHANGELOG.md (hunk 3 above)
Comment (thomaseizinger) [reply swarm/CHANGELOG.md]: Sounds good to me!

File: swarm/src/handler.rs (hunk 2 above)
Comment (thomaseizinger) [reply swarm/src/handler.rs:472]: Including this information will be a breaking change but we can introduce the facilities for it in a non-breaking manner once we have completely encapsulated `multistream-select`, meaning this is a step in the right direction in my opinion.

**Pull request merged**